### PR TITLE
Required fixes for future Groovy 4 compatibility

### PR DESCRIPTION
**Pull request by DPUkyle**
Extracted from #20038

### CODE REVIEW

**Comment by DPUkyle** [comment]
@bot-gradle test this

File: subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/codenarc/CodeNarcPluginVersionIntegrationTest.groovy [hunk 1]
```diff
@@ -48,7 +51,7 @@ class CodeNarcPluginVersionIntegrationTest extends MultiVersionIntegrationSpec {
             ${JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_14) ?
             """
             configurations.codenarc {
-                resolutionStrategy.force 'org.codehaus.groovy:groovy:${GroovySystem.version}'
+                resolutionStrategy.force 'org.codehaus.groovy:groovy:3.0.11' // force latest Groovy 3 when using Java 14+.  Do not use GroovySystem#version as Groovy 4 needs different coordinates
```
Comment (octylFractal) [inline subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/codenarc/CodeNarcPluginVersionIntegrationTest.groovy:54]: Is there a reason you didn't use `3.0.12` here? Or even better, `GroovyCoverage.MINIMAL_GROOVY_3`?

File: subprojects/launcher/src/test/groovy/org/gradle/launcher/cli/WelcomeMessageActionTest.groovy [hunk 2]
```diff
@@ -114,7 +114,7 @@ For more details see https://docs.gradle.org/42.0/release-notes.html''')
         given:
         def inputStreamWasClosed = false
         def inputStreamProvider = Mock(Function) {
-            apply(_) >> new ByteArrayInputStream("".bytes) {
+            apply(_) >> new ByteArrayInputStream(Byte.parseByte('0')) {
```
Comment (octylFractal) [inline subprojects/launcher/src/test/groovy/org/gradle/launcher/cli/WelcomeMessageActionTest.groovy:117]: What exactly goes wrong here in Groovy 4? The original code seems to work okay in groovysh.

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/JavaPropertyReflectionUtilTest.groovy [hunk 3]
```diff
@@ -355,7 +355,7 @@ class WithProperties {
 
     void setWriteOnly(String s1) { }
 
-    Boolean isSomething() { return null }
+    boolean isSomething() { return null }
```
Comment (octylFractal) [inline subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/JavaPropertyReflectionUtilTest.groovy:358]: Does this work? This seems odd.

File: subprojects/reporting/src/integTest/groovy/org/gradle/api/reporting/plugins/BuildDashboardPluginIntegrationTest.groovy [hunk 4]
```diff
@@ -79,7 +81,7 @@ class BuildDashboardPluginIntegrationTest extends WellBehavedPluginTest {
                 }
             }
 
-            ${JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_14) ?
+            ${ JavaVersion.current().isCompatibleWith(JavaVersion.VERSION_14) ?
```
Comment (octylFractal) [inline subprojects/reporting/src/integTest/groovy/org/gradle/api/reporting/plugins/BuildDashboardPluginIntegrationTest.groovy:84]: Unnecessary change.

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy [hunk 5]
```diff
@@ -16,11 +16,15 @@
...
 
+import org.gradle.util.internal.VersionNumber
+import org.junit.Assume
+
 import static org.gradle.testkit.runner.TaskOutcome.SUCCESS
 
 class GradleVersionsPluginSmokeTest extends AbstractPluginValidatingSmokeTest {
 
     def 'can check for updated versions'() {
+        Assume.assumeTrue('Incompatible with Groovy 4: Execution failed for task \':dependencyUpdates\'. > groovy/util/XmlSlurper', VersionNumber.parse(GroovySystem.version).major < 4) // TODO Watch for merge and release of https://github.com/ben-manes/gradle-versions-plugin/pull/656
```
Comment (octylFractal) [inline subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy:27]: Perhaps we should comment on this PR? It seems dead.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testsuites/TestSuitesIntegrationTest.groovy [hunk 6]
```diff
@@ -301,17 +301,18 @@ class TestSuitesIntegrationTest extends AbstractIntegrationSpec {
...
-        'useJUnitJupiter("5.7.1")'   | JUnitPlatformTestFramework | "junit-jupiter-5.7.1.jar"
-        'useSpock()'                 | JUnitPlatformTestFramework | "spock-core-${DefaultJvmTestSuite.Frameworks.SPOCK.getDefaultVersion()}.jar"
-        'useSpock("2.1-groovy-3.0")' | JUnitPlatformTestFramework | "spock-core-2.1-groovy-3.0.jar" // Not possible to test a different version from the default yet, since this is the first groovy 3.0 targeted release
-        'useKotlinTest()'            | JUnitPlatformTestFramework | "kotlin-test-junit5-${DefaultJvmTestSuite.Frameworks.KOTLIN_TEST.getDefaultVersion()}.jar"
-        'useKotlinTest("1.5.30")'    | JUnitPlatformTestFramework | "kotlin-test-junit5-1.5.30.jar"
-        'useTestNG()'                | TestNGTestFramework        | "testng-${DefaultJvmTestSuite.Frameworks.TESTNG.getDefaultVersion()}.jar"
-        'useTestNG("7.3.0")'         | TestNGTestFramework        | "testng-7.3.0.jar"
+        testingFrameworkDeclaration     | testingFrameworkType       | testingFrameworkDep
```
Comment (octylFractal) [inline subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testsuites/TestSuitesIntegrationTest.groovy:304]: I am not sure why this was reformatted.

**Review by octylFractal — CHANGES_REQUESTED** [review]

File: subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy (hunk 5 above)
Comment (DPUkyle) [reply subprojects/smoke-test/src/smokeTest/groovy/org/gradle/smoketests/GradleVersionsPluginSmokeTest.groovy:27]: [Done](https://github.com/ben-manes/gradle-versions-plugin/pull/656#issuecomment-1246089846)

File: subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/codenarc/CodeNarcPluginVersionIntegrationTest.groovy (hunk 1 above)
Comment (DPUkyle) [reply subprojects/code-quality/src/integTest/groovy/org/gradle/api/plugins/quality/codenarc/CodeNarcPluginVersionIntegrationTest.groovy:54]: Good idea. At the time, 3.0.11 was the newest.

File: subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/JavaPropertyReflectionUtilTest.groovy (hunk 3 above)
Comment (DPUkyle) [reply subprojects/model-core/src/test/groovy/org/gradle/internal/reflect/JavaPropertyReflectionUtilTest.groovy:358]: You're right, this is a nonsense example. I'll change the implementation to return false. 

This change was originally done as Groovy 4 doesn't support wrapped Booleans as properties, causing this test to fail.

File: subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testsuites/TestSuitesIntegrationTest.groovy (hunk 6 above)
Comment (DPUkyle) [reply subprojects/testing-jvm/src/integTest/groovy/org/gradle/testing/testsuites/TestSuitesIntegrationTest.groovy]: Thank you; fixed

File: subprojects/reporting/src/integTest/groovy/org/gradle/api/reporting/plugins/BuildDashboardPluginIntegrationTest.groovy (hunk 4 above)
Comment (DPUkyle) [reply subprojects/reporting/src/integTest/groovy/org/gradle/api/reporting/plugins/BuildDashboardPluginIntegrationTest.groovy]: Thank you; fixed

**Comment by DPUkyle** [comment]
@bot-gradle test this

**Review by octylFractal — APPROVED** [review]

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

File: subprojects/launcher/src/test/groovy/org/gradle/launcher/cli/WelcomeMessageActionTest.groovy (hunk 2 above)
Comment (octylFractal) [reply subprojects/launcher/src/test/groovy/org/gradle/launcher/cli/WelcomeMessageActionTest.groovy:117]: Weird. I don't understand what would cause that. It might be better to just use `(byte) 0` here though.

Comment (DPUkyle) [reply subprojects/launcher/src/test/groovy/org/gradle/launcher/cli/WelcomeMessageActionTest.groovy:117]: What's _even stranger_ is that the BAIS ctor accepts a byte array, not a byte. Maybe something is wrong with spock's mocking?

Here's the short ST when I tried passing `new byte[0]`:

```
org.codehaus.groovy.runtime.typehandling.GroovyCastException: Cannot cast object '[B@5f8d9767' with class '[B' to class 'byte'
	at org.gradle.launcher.cli.WelcomeMessageActionTest.closes InputStream after reading features_closure3(WelcomeMessageActionTest.groovy:117)
	at groovy.lang.Closure.call(Closure.java:418)
	at org.spockframework.lang.SpecInternals.createMock(SpecInternals.java:53)
	at org.spockframework.lang.SpecInternals.createMockImpl(SpecInternals.java:305)
	at org.spockframework.lang.SpecInternals.createMockImpl(SpecInternals.java:295)
	at org.spockframework.lang.SpecInternals.MockImpl(SpecInternals.java:108)
	at org.gradle.launcher.cli.WelcomeMessageActionTest.closes InputStream after reading features(WelcomeMessageActionTest.groovy:116)
```

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Comment by DPUkyle** [comment]
@bot-gradle test and merge

**Comment by octylFractal** [comment]
@bot-gradle test and merge
